crates/farming/src/lib.rs

```diff
@@ -0,0 +1,357 @@
+                    .is_zero(),
+                Error::<T>::ScheduleNotFound
+            );
+
+            // reserve lp tokens to prevent spending
+            T::MultiCurrency::reserve(pool_currency_id.clone(), &who, amount)?;
+
+            // deposit lp tokens as stake
+            T::RewardPools::deposit_stake(&pool_currency_id, &who, amount)
+        }
+
+        /// Unstake the pool tokens from the reward pool
+        #[pallet::weight(T::WeightInfo::withdraw())]
+        #[transactional]
+        pub fn withdraw(
+            origin: OriginFor<T>,
+            pool_currency_id: CurrencyIdOf<T>,
+            amount: BalanceOf<T>,
+        ) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+
+            // unreserve lp tokens to allow spending
+            let _remaining = T::MultiCurrency::unreserve(pool_currency_id.clone(), &who, amount);
+            // TODO: check remaining is non-zeo
+
+            // withdraw lp tokens from stake
+            T::RewardPools::withdraw_stake(&pool_currency_id, &who, amount)
+        }
+
+        /// Withdraw any accrued rewards from the reward pool
+        #[pallet::weight(T::WeightInfo::claim())]
+        #[transactional]
+        pub fn claim(
+            origin: OriginFor<T>,
+            pool_currency_id: CurrencyIdOf<T>,
+            reward_currency_id: CurrencyIdOf<T>,
+        ) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+            let pool_account_id = Self::pool_account_id(&pool_currency_id);
+
+            // get reward from staking pool
+            let reward = T::RewardPools::withdraw_reward(&pool_currency_id, &who, reward_currency_id)?;
+            // transfer from pool to user
+            T::MultiCurrency::transfer(reward_currency_id, &pool_account_id, &who, reward)?;
+
+            Self::deposit_event(Event::RewardClaimed {
+                account_id: who,
+                pool_currency_id,
+                reward_currency_id,
+                amount: reward,
+            });
+
+            Ok(())
+        }
+    }
+}
+
+// "Internal" functions, callable by code.
+impl<T: Config> Pallet<T> {
+    pub fn pool_account_id(pool_currency_id: &CurrencyIdOf<T>) -> T::AccountId {
+        T::FarmingPalletId::get().into_sub_account_truncating(pool_currency_id)
+    }
+
+    pub fn treasury_account_id() -> T::AccountId {
+        T::TreasuryPalletId::get().into_account_truncating()
+    }
+
+    fn try_distribute_reward(
+        pool_currency_id: CurrencyIdOf<T>,
+        reward_currency_id: CurrencyIdOf<T>,
+        amount: BalanceOf<T>,
+    ) -> Result<(), DispatchError> {
+        frame_support::storage::with_transaction::<_, DispatchError, _>(|| {
+            let res = T::RewardPools::distribute_reward(&pool_currency_id, reward_currency_id, amount);
+            if res.is_ok() {
+                TransactionOutcome::Commit(res)
+            } else {
+                TransactionOutcome::Rollback(res)
+            }
+        })
+    }
```

Comment:
You should be able to do something like this
```suggestion
    #[transactional]
    fn try_distribute_reward(
        pool_currency_id: CurrencyIdOf<T>,
        reward_currency_id: CurrencyIdOf<T>,
        amount: BalanceOf<T>,
    ) -> Result<(), DispatchError> {
        T::RewardPools::distribute_reward(&pool_currency_id, reward_currency_id, amount)
    }
```